src/features/targeting/useMessage.js

```diff
@@ -1,36 +1,48 @@
 // @flow
-import { useContext } from 'react';
+import { useContext, useMemo } from 'react';
 import { MESSAGE_STATES } from './constants';
-import type { MessageName } from './types';
-import type { TargetingApi } from './TargetingApi';
+import type { MessageName, TargetingApi } from './types';
 import MessageContext from './MessageContext';
 
 function useMessage(name: MessageName): TargetingApi {
-    const {
-        messageApi: { eligibleMessageIDMap, markMessageAsSeen, markMessageAsClosed },
-        messageStateMap,
-        setMessageStateMap,
-    } = useContext(MessageContext);
+    const messageContext = useContext(MessageContext);
+    return useMemo(() => {
+        const {
+            messageApi: { eligibleMessageIDMap, markMessageAsSeen, markMessageAsClosed },
+            messageStateMap,
+            setMessageStateMap,
+        } = messageContext;
 
-    function canShow() {
-        return name in eligibleMessageIDMap && messageStateMap[name] !== MESSAGE_STATES.CLOSED;
-    }
+        const canShow = name in eligibleMessageIDMap && messageStateMap[name] !== MESSAGE_STATES.CLOSED;
 
-    function onShow() {
-        if (canShow() && !messageStateMap[name]) {
-            setMessageStateMap({ ...messageStateMap, [name]: MESSAGE_STATES.SHOWING });
-            markMessageAsSeen(eligibleMessageIDMap[name]);
+        function onShow() {
+            if (canShow && !messageStateMap[name]) {
+                setMessageStateMap({ ...messageStateMap, [name]: MESSAGE_STATES.SHOWING });
+                // FIXME markMessageAsSeen action was currently throttled to prevent we make multiple
+                // backend call if onShow is called multiple times before setMessageStateMap
+                // actually alter the messageState. But it is preferrable to prevent
+                // markMessageAsSeen from being called multiple times instead of throttling
+                // after it is called.
```

Comment:
Is this something that can be fixed once we implement more of the state machine?

Reply:
Yeah, actually the useShowOne hook will alleviate this so its not a big deal for now but we should figure out a better way eventually

Reply:
hi, we internally discussed this. The "shown" state should be a closure state but not react state. Once we fix that way, the bugs will be all gone @JohnnyB83 , however, it is not urgent.